x/protorev/keeper/hooks.go

```diff
@@ -0,0 +1,215 @@
+
+// afterPoolCreatedWithCoins checks if the new pool should be stored as the highest liquidity pool
+// for any of the base denoms, and stores it if so.
+func (k Keeper) afterPoolCreatedWithCoins(ctx sdk.Context, poolId uint64) {
+	baseDenoms, err := k.GetAllBaseDenoms(ctx)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting base denoms in AfterCFMMPoolCreated hook", err)
+		return
+	}
+
+	baseDenomMap := make(map[string]bool)
+	for _, baseDenom := range baseDenoms {
+		baseDenomMap[baseDenom.Denom] = true
+	}
+
+	pool, err := k.poolmanagerKeeper.GetPool(ctx, poolId)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting pool in AfterCFMMPoolCreated hook", err)
+		return
+	}
+
+	denoms, err := k.poolmanagerKeeper.RouteGetPoolDenoms(ctx, poolId)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting pool liquidity in afterPoolCreated", err)
+		return
+	}
+
+	// Pool must be active and the number of denoms must be 2
+	if pool.IsActive(ctx) && len(denoms) == 2 {
+		if baseDenomMap[denoms[0]] {
+			k.compareAndStorePool(ctx, poolId, denoms[0], denoms[1])
+		}
+		if baseDenomMap[denoms[1]] {
+			k.compareAndStorePool(ctx, poolId, denoms[1], denoms[0])
+		}
+	}
+}
+
+// storeSwap stores a swap to be checked by protorev when attempting backruns.
+func (k Keeper) storeSwap(ctx sdk.Context, poolId uint64, tokenIn, tokenOut string) {
+	swapToBackrun := types.Trade{
+		Pool:     poolId,
+		TokenIn:  tokenIn,
+		TokenOut: tokenOut,
+	}
+
+	if err := k.AddSwapsToSwapsToBackrun(ctx, []types.Trade{swapToBackrun}); err != nil {
+		ctx.Logger().Error("Protorev error adding swap to backrun from storeSwap", err) // Does not return since logging is last thing in the function
+	}
+}
+
+// compareAndStorePool compares the liquidity of the new pool with the liquidity of the stored pool, and stores the new pool if it has more liquidity.
+func (k Keeper) compareAndStorePool(ctx sdk.Context, poolId uint64, baseDenom, otherDenom string) {
+	storedPoolId, err := k.GetPoolForDenomPair(ctx, baseDenom, otherDenom)
+	if err != nil {
+		// Error means no pool exists for this pair, so we set it
+		k.SetPoolForDenomPair(ctx, baseDenom, otherDenom, poolId)
+		return
+	}
+
+	defer func() {
+		if r := recover(); r != nil {
+			ctx.Logger().Error("Protorev error recovering from panic in AfterCFMMPoolCreated hook, likely an overflow error", r)
+		}
+	}()
+
+	// Get coins and calculate liquidity for the new pool
+	newPoolCoins, err := k.poolmanagerKeeper.GetTotalPoolLiquidity(ctx, poolId)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting newPoolCoins in compareAndStorePool", err)
+		return
+	}
+	newPoolliquidity := newPoolCoins[0].Amount.Mul(newPoolCoins[1].Amount)
+
+	// Get coins and calculate liquidity for the stored pool
+	storedPool, err := k.gammKeeper.GetPoolAndPoke(ctx, storedPoolId)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting storedPool in AfterCFMMPoolCreated hook", err)
+		return
+	}
```

Comment:
This only gets balancer pools, what about CL?

Reply:
Nice catch, this needs to use poolManager instead

Reply:
Addressed here: https://github.com/osmosis-labs/osmosis/pull/5045/commits/91f16f0e0bc60638d002c21a7a2224e8fdd5bf72